lib/avro/builder/type_factory.rb

```diff
@@ -4,28 +4,33 @@ module Builder
     # This concern is used by classes that create new Type instances.
     module TypeFactory
 
+      SPECIFIC_TYPES = Set.new(%w(Array Enum Fixed Map Record Union).map(&:freeze)).freeze
+
       private
 
-      # Return a new Type instance
+      # Return a new Type instance or nil
       def create_type(type_name)
         case
         when Avro::Schema::PRIMITIVE_TYPES_SYM.include?(type_name.to_sym)
           Avro::Builder::Types::Type.new(type_name)
         else
-          type_class_name = "#{type_name.to_s.capitalize}Type"
-          Avro::Builder::Types.const_get(type_class_name).new
+          type_class_name = "#{type_name.to_s.capitalize}"
+          if SPECIFIC_TYPES.include?(type_class_name)
+            Avro::Builder::Types.const_get("#{type_class_name}Type").new
+          end
         end
       end
 
       # Return a new Type instance, including propagating internal state
       # and setting attributes via the DSL
       def build_type(type_name, field: nil, builder: nil, internal: {}, options: {}, &block)
```

Comment:
`build_type` calls `create_type` to create a new instance and then configures that new instance.